### PR TITLE
feat(repo): support search repository by topic name

### LINKED ISSUE #4504: Can't search repository by topic name
- Gitea version (or commit ref): ae9dd239fbfc3b9f18ff2ccbdbe6fdf236db2b36
- Git version: 2.17.1
- Operating system:
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL)
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description

According to https://github.com/go-gitea/gitea/pull/3711#issuecomment-406674344 comments. User can't search repository by topic name.

### PR DESCRIPTION
**Pull request by appleboy**
fix #4504 

admin46_test_-_bo-yi_wu

### CODE REVIEW

**Comment by gsantner** [comment]
This will also find repositories that are not related to the topic/tag. E.g. if all go-repos get tagged with go, and I click on the tag I also get repos that have only go somewhere in text, but are not tagged with go

**Comment by appleboy** [comment]
@gsantner Fixed. Add the new bool flag `topic`.

**Comment by lafriks** [comment]
I think it should be different argument for topics so that it would be possible filter by both name and topics at the same time

**Comment by appleboy** [comment]
@lafriks Default behavior is filtered by both name and topics and only search by topic if add `topic=1` in query string. ok?

**Comment by appleboy** [comment]
Testing Pass finally :)

**Comment by appleboy** [comment]
@gsantner @gsantner any feedbacks?

**Comment by gsantner** [comment]
I'm fine if I get the results (and only the results) for given topic, which its the case :+1: :D

File: models/repo_list.go [hunk 1]
```diff
@@ -202,7 +204,14 @@ func SearchRepositoryByName(opts *SearchRepoOptions) (RepositoryList, int64, err
 	}
 
 	if opts.Keyword != "" {
-		cond = cond.And(builder.Like{"lower_name", strings.ToLower(opts.Keyword)})
+		var keywordCond = builder.NewCond()
+		if opts.Topic {
```
Comment (lunny) [inline models/repo_list.go]:
```Go
keywordCond = keywordCond.Or(builder.Like{"topic.name", strings.ToLower(opts.Keyword)})
if !opts.TopicOnly {
    keywordCond = keywordCond.Or(builder.Like{"lower_name", strings.ToLower(opts.Keyword)})
}

**Comment by lunny** [comment]
@appleboy I think rename `Topic` to `TopicOnly` is better.

**Comment by lunny** [comment]
LGTM

**Comment by appleboy** [comment]
@lafriks Waiting for your feedback

**Comment by appleboy** [comment]
@lafriks any feedbacks?

**Comment by gsantner** [comment]
@appleboy  I guess first response will be: squash commits :D

**Comment by lafriks** [comment]
Yes I still think topics should be provided as different argument to url preferably also by splitting  provided values by `,` so that multiple topics could be provided

**Comment by appleboy** [comment]
@lafriks maybe I can create another PR to implement this feature for multiple topics search.

**Comment by techknowlogick** [comment]
LGTM, please resolve conflicting files

**Pull request merged**